data/oonireport.1
.\" Man page generated from reStructuredText.
.
.TH "OONIREPORT" "1" "Feb 14, 2018" "2.3.0" "OONI: Open Observatory of Network Interference"
.SH NAME
oonireport \- upload ooniprobe measurements
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.sp
\fBoonireport\fP [\fIoptions\fP] upload | status [\fIpath to report\fP]
.SH DESCRIPTION
.sp
\fBoonireport\fP is a tool for viewing which ooniprobe
reports have not been submitted to a collector and upload them.
.sp
\fBooniprobe \-i /usr/share/ooni/decks\-available/web.yaml\fP
.SH OPTIONS
.INDENT 0.0
.TP
.B \-d\fP,\fB  \-\-default\-collector
Upload the reports to the default collector that is looked up
with the canonical bouncer.
.TP
.B \-f\fP,\fB  \-\-configfile
Specify the configuration file to use.
.TP
.B \-c\fP,\fB  \-\-collector
Specify the collector to upload the result to.
.TP
.B \-b\fP,\fB  \-\-bouncer
Specify the bouncer to query for a collector.
.sp
–version
.INDENT 7.0
.TP
.B \-\-help
Display this help and exit.
.UNINDENT
.UNINDENT
.SH COMMANDS
.sp
\fBupload\fP
.sp
If no argument is specified all the reports that have not been
submitted will be submitted to either the collector specified as
an option (\-c) or to the collector that was previously used.
.sp
If a report file is specified then only that report file will be
uploaded
.sp
\fBstatus\fP
.sp
Outputs all of the reports that are either “not uploaded”,
“incomplete” or “pending”.
.sp
The possible statuses are:
.sp
not\-created
.sp
If it has not been created, because the user specified to not use a
collector.
.sp
creation\-failed
.sp
If we attempted to create the report, but failed to do so either because
the collector does not accept our report or because it was unreachable at the
time.
.sp
created
.sp
If the report has been successfully created, but has not yet been
closed.
.sp
incomplete
.sp
If the report has been created, but we have failed to update the
report with some data.
.SH OONIREPORT
.sp
This tool is used to upload reports that are stored on the users
filesystem and have not been submitted to a collector. This can be
either because the collector was unreachable when we attempted to
submit the report or because the policy of the collector did not
support the specified test.
.SH AUTHOR
The Tor Project
.SH COPYRIGHT
2016, The Tor Project
.\" Generated by docutils manpage writer.
.


ChangeLog.md
## v1.1.0

In this new release of ooniprobe we have added a new command line tool for
listing the reports that have not been published to a collector and that allows
the probe operator to choose which ones they would like to upload.

We have also made some privacy improvements to the reports (we will sanitize
all things that may look like file paths) and added metadata associated with
the maxmind database being used by the probe operator.

Here is a more detailed list of what has been done:

* Annotate on disk which reports we have submitted and which ones we have not:
  https://trac.torproject.org/projects/tor/ticket/11860

* Add tool called oonireport for publishing unpublished ooniprobe reports to a
  collector: https://trac.torproject.org/projects/tor/ticket/11862

* Probe Report does not leak filepaths anymore:
  https://trac.torproject.org/projects/tor/ticket/12706

* Reports now include version information about the maxmind database being
  used: https://trac.torproject.org/projects/tor/ticket/12771

* We now do integrity checks on the ooniprobe.conf file so that we don't start
  the tool if the config file is missing some settings or is not consistent:
  https://trac.torproject.org/projects/tor/ticket/11983
  (thanks to Alejandro López (kudrom))

* Improvements have been made to the sniffer subsystem (thanks to Alejandro
  López (kudrom))

Minor bug fixes:

* Fix dns_spoof test (by kudrom)
  https://trac.torproject.org/projects/tor/ticket/12486

* ooni might not look at requiresTor:
  https://trac.torproject.org/projects/tor/ticket/11858

* ooni spits out gobs of tracebacks if Tor is not running and the OONI config
  says it will be:
  https://trac.torproject.org/projects/tor/ticket/11859

* The README for ooni-probe should mention the bugtracker and repository
  https://trac.torproject.org/projects/tor/ticket/11980

## v1.0.2

* Add ooniprobe manpage.

* Fix various security issues raised by the least authority audit.

* Add a test that checks for Tor bridge reachability.

* Record the IP address of the exit node being used in torified requests.

* Captive portal test now uses the ooni-probe test templates.

* Have better test naming consistency.

## v1.0.1

* Fix bugs in the traceroute test that lead to not all packets being collected.

* All values inside of http_requests test are now initialized inside of setUp.

* Fix a bug that lead to the input value of the report not being set in some
  circumstances.

* Add bridge_reachability test

## v1.0.0

* Add bouncer support for discovering test helpers and collectors

* Fix bug that lead to HTTP tests to stall

* Add support for connect_error and connection_lost_error error types

* Add support for additional Tor configuration keys

* Add disclaimer when running ooniprobe

## v0.1.0

Improvements to HTML/JS based user interface:

  * XSRF protection

  * user supplied input specification

Bugfixing and improvements to scheduler.

## v0.0.12

Implement JS/HTML based user interface.

Supports:

  * Starting and stopping of tests

  * Monitoring of test progress

## v0.0.11

* Parametrize task timeout and retry count

* Set the default collector via the command line option

* Add option to disable the default collector

* Add continuous integration with travis

## v0.0.10

### ooniprobe

* Fix bug that made HTTP based tests stall

* Update DNS Test example to not import the DNS Test template If you import the
	DNS Test template it will be considered a valid test case and command line
	argument parsing will not work as expected. see:
	#7795 for more details

* Fix major bug in DNS test template that prevented PTR lookups from working
	properly I was calling the queryUDP function with the arguments in the wrong
	order. Twisted, why you API no consistent?

* Add support for specifying the level of parallelism in tests (aka router
	melt mode)

* Do not swallow failures when a test instance fails to run fixes #7714

### scripts

* Add report archival script

Fix bug in TCP connect test that made it not properly log errors

* Refactor failure handling code in nettest Add function that traps all the
	supported failure and outputs the failure string representing it.

### documentation

* Add birdseye view of the ooniprobe architecture

* Add details on the current implementation status of ooni*

* Add draft ooniprobe API specification

* Add instructions for supervisord configuration and clean up README.md

## 0.0.9

### ooniprobe

* Set the default ASN to 0

* Make Beautiful soup a soft depedency

* Add support for sending the ASN number of the probe:
	the ASN number will get sent when creating a new report

* Add support for obtaining the probes IP address via getinfo address as per
	https://trac.torproject.org/projects/tor/ticket/7447

* Fix bug in ooniprobe test decks
	https://trac.torproject.org/projects/tor/ticket/7664

### oonib

* Use twisted fdesc when writing to files

* Add support for processing the ASN number of the probe

* Test reports shall follow the specification detailed inside of docs/reports.rst

* Add support for setting the tor binary path in oonib/config.py

### scripts

* Add a very simple example on how to securely parse the ooniprobe reports

### documentation

* Add documentation for the DNSSpoof test

* Add documentation for HTTPHeaderFieldManipulation

* Clean up writing_tests.rst

* Properly use the power of sphinx!

### Tests

* fixup Netalyzr third party plugin

## v0.0.8-alpha

### ooniprobe

* Allow test resolver file to have comments.

* Autostart Tor in default configuration.

* Add support for starting Tor via txtorcon.

* Make the sniffer not run in a separate thread, but use a non blocking fdesc.
	Do some refactoring of scapy testing, following Factory creational pattern
	and a pub-sub pattern for the readers and writers.

* Extend TrueHeaders to support calculation of difference between two HTTP headers respectful of
	capitalization


ooni/scripts/oonireport.py
from __future__ import print_function

import os
import sys
import json
import yaml

from twisted.python import usage
from twisted.internet import defer, task, reactor

from ooni.constants import CANONICAL_BOUNCER_ONION
from ooni.reporter import OONIBReporter, OONIBReportLog

from ooni.utils import log
from ooni.settings import config
from ooni.backend_client import BouncerClient, CollectorClient
from ooni import __version__

@defer.inlineCallbacks
def lookup_collector_client(report_header, bouncer):
    oonib_client = BouncerClient(bouncer)
    net_tests = [{
        'test-helpers': [],
        'input-hashes': [],
        'name': report_header['test_name'],
        'version': report_header['test_version'],
    }]
    result = yield oonib_client.lookupTestCollector(
        net_tests
    )
    collector_client = CollectorClient(
        address=result['net-tests'][0]['collector']
    )
    defer.returnValue(collector_client)

class NoIDFound(Exception):
    pass

def report_path_to_id(report_file):
    measurement_dir = os.path.dirname(report_file)
    measurement_id = os.path.basename(measurement_dir)
    if os.path.dirname(measurement_dir) != config.measurements_directory:
        raise NoIDFound
    return measurement_id

@defer.inlineCallbacks
def upload(report_file, collector=None, bouncer=None, measurement_id=None):
    oonib_report_log = OONIBReportLog()
    collector_client = None
    if collector:
        collector_client = CollectorClient(address=collector)

    try:
        # Try to guess the measurement_id from the file path
        measurement_id = report_path_to_id(report_file)
    except NoIDFound:
        pass

    log.msg("Attempting to upload %s" % report_file)

    if report_file.endswith(".njson"):
        report = NJSONReportLoader(report_file)
    else:
        log.warn("Uploading of YAML formatted reports will be dropped in "
                 "future versions")
        report = YAMLReportLoader(report_file)

    if bouncer and collector_client is None:
        collector_client = yield lookup_collector_client(report.header,
                                                         bouncer)

    if collector_client is None:
        if measurement_id:
            report_log = yield oonib_report_log.get_report_log(measurement_id)
            collector_settings = report_log['collector']
            print(collector_settings)
            if collector_settings is None or len(collector_settings) == 0:
                log.warn("Skipping uploading of %s since this measurement "
                         "was run by specifying no collector." %
                          report_file)
                defer.returnValue(None)
            elif isinstance(collector_settings, dict):
                collector_client = CollectorClient(settings=collector_settings)
            elif isinstance(collector_settings, str):
                collector_client = CollectorClient(address=collector_settings)
        else:
            log.msg("Looking up collector with canonical bouncer." % report_file)
            collector_client = yield lookup_collector_client(report.header,
                                                             CANONICAL_BOUNCER_ONION)

    oonib_reporter = OONIBReporter(report.header, collector_client)
    log.msg("Creating report for %s with %s" % (report_file,
                                                collector_client.settings))
    report_id = yield oonib_reporter.createReport()
    report.header['report_id'] = report_id
    if measurement_id:
        log.debug("Marking it as created")
        yield oonib_report_log.created(measurement_id,
                                       collector_client.settings)
    log.msg("Writing report entries")
    for entry in report:
        yield oonib_reporter.writeReportEntry(entry)
        log.msg("Written entry")
    log.msg("Closing report")
    yield oonib_reporter.finish()
    if measurement_id:
        log.debug("Closing log")
        yield oonib_report_log.closed(measurement_id)


@defer.inlineCallbacks
def upload_all(collector=None, bouncer=None, upload_incomplete=False):
    oonib_report_log = OONIBReportLog()

    reports_to_upload = yield oonib_report_log.get_to_upload()
    for report_file, value in reports_to_upload:
        try:
            yield upload(report_file, collector, bouncer,
                         value['measurement_id'])
        except Exception as exc:
            log.exception(exc)

    if upload_incomplete:
        reports_to_upload = yield oonib_report_log.get_incomplete()
        for report_file, value in reports_to_upload:
            try:
                yield upload(report_file, collector, bouncer,
                             value['measurement_id'])
            except Exception as exc:
                log.exception(exc)

def print_report(report_file, value):
    print("* %s" % report_file)
    print("  %s" % value['last_update'])


@defer.inlineCallbacks
def status():
    oonib_report_log = OONIBReportLog()

    reports_to_upload = yield oonib_report_log.get_to_upload()
    print("Reports to be uploaded")
    print("----------------------")
    for report_file, value in reports_to_upload:
        print_report(report_file, value)

    reports_in_progress = yield oonib_report_log.get_in_progress()
    print("Reports in progress")
    print("-------------------")
    for report_file, value in reports_in_progress:
        print_report(report_file, value)

    reports_incomplete = yield oonib_report_log.get_incomplete()
    print("Incomplete reports")
    print("------------------")
    for report_file, value in reports_incomplete:
        print_report(report_file, value)

class ReportLoader(object):
    _header_keys = (
        'probe_asn',
        'probe_cc',
        'probe_ip',
        'probe_city',
        'test_start_time',
        'test_name',
        'test_version',
        'options',
        'input_hashes',
        'software_name',
        'software_version',
        'data_format_version',
        'report_id',
        'test_helpers',
        'annotations',
        'id'
    )

    def __iter__(self):
        return self

    def close(self):
        self._fp.close()

class YAMLReportLoader(ReportLoader):
    def __init__(self, report_filename):
        self._fp = open(report_filename)
        self._yfp = yaml.safe_load_all(self._fp)

        self.header = self._yfp.next()

    def next(self):
        try:
            return self._yfp.next()
        except StopIteration:
            self.close()
            raise StopIteration

class NJSONReportLoader(ReportLoader):
    def __init__(self, report_filename):
        self._fp = open(report_filename)
        self.header = self._peek_header()

    def _peek_header(self):
        header = {}
        first_entry = json.loads(next(self._fp))
        for key in self._header_keys:
            header[key] = first_entry.get(key, None)
        self._fp.seek(0)
        return header

    def next(self):
        try:
            entry = json.loads(next(self._fp))
            for key in self._header_keys:
                entry.pop(key, None)
            test_keys = entry.pop('test_keys')
            entry.update(test_keys)
            return entry
        except StopIteration:
            self.close()
            raise StopIteration

class Options(usage.Options):

    synopsis = """%s [options] upload | status
""" % (os.path.basename(sys.argv[0]),)

    optFlags = [
        ["default-collector", "d", "Upload the reports to the default "
                                   "collector that is looked up with the "
                                   "canonical bouncer."]
    ]

    optParameters = [
        ["configfile", "f", None,
         "Specify the configuration file to use."],
        ["collector", "c", None,
         "Specify the collector to upload the result to."],
        ["bouncer", "b", None,
         "Specify the bouncer to query for a collector."]
    ]

    def opt_version(self):
        print("oonireport version: %s" % __version__)
        sys.exit(0)

    def parseArgs(self, *args):
        if len(args) == 0:
            raise usage.UsageError(
                "Must specify at least one command"
            )
            return
        self['command'] = args[0]
        if self['command'] not in ("upload", "status"):
            raise usage.UsageError(
                "Must specify either command upload or status"
            )
        if self['command'] == "upload":
            try:
                self['report_file'] = args[1]
            except IndexError:
                self['report_file'] = None


def tor_check():
    if not config.tor.socks_port:
        log.err("Currently oonireport requires that you start Tor yourself "
                "and set the socks_port inside of ooniprobe.conf")
        sys.exit(1)


def oonireport(_reactor=reactor, _args=sys.argv[1:]):
    options = Options()
    try:
        options.parseOptions(_args)
    except Exception as exc:
        print("Error: %s" % exc)
        print(options)
        sys.exit(2)
    config.global_options = dict(options)
    config.set_paths()
    config.read_config_file()

    if options['default-collector']:
        options['bouncer'] = CANONICAL_BOUNCER_ONION

    if options['command'] == "upload" and options['report_file']:
        log.start()
        tor_check()
        return upload(options['report_file'],
                      options['collector'],
                      options['bouncer'])
    elif options['command'] == "upload":
        log.start()
        tor_check()
        return upload_all(options['collector'],
                          options['bouncer'])
    elif options['command'] == "status":
        return status()
    else:
        print(options)

def run():
    task.react(oonireport)

if __name__ == "__main__":
    run()
